### PR TITLE
Implement file-of-patterns functionality

### PR DESCRIPTION
**Pull request by tneymanov**
Allows users to specify a list of files or input pattern inside a file to ingest into the tool.

Notes:
- Could be helpful for cases when there isn't an easy regex for users to select all of their files.
- VCF to BQ preprocessor is also modified to accept both inputs.
- Using input_file flag method automatically quasi-turns on optimize_for_large_inputs, so it would be preferable not to use it for inputs of less than 50k files.

Issue: [441](https://github.com/googlegenomics/gcp-variant-transforms/issues/441)

### CODE REVIEW

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_move_to_call_from_file.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+[
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_move_to_call_from_file.json]: Done, running the tests right now.

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/combine_from_multiple_inputs.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+[
+  {
+    "test_name": "combine-from-multiple-inputs",
+    "table_name": "combine_from_multiple_inputs",
+    "input_file": "gs://gcp-variant-transforms-testfiles/small_tests/input_files/combine_input",
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/combine_from_multiple_inputs.json]: Done.

File: gcp_variant_transforms/pipeline_common.py [hunk 3]
```diff
@@ -69,33 +69,80 @@ def parse_args(argv, command_line_options):
   return known_args, pipeline_args
 
 
-def get_pipeline_mode(input_pattern, optimize_for_large_inputs=False):
+def get_pipeline_mode(
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common.py]: Done.

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 4]
```diff
@@ -26,7 +26,6 @@
 from gcp_variant_transforms.beam_io import vcfio
 from gcp_variant_transforms.libs import bigquery_sanitizer
 
-
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: I thought python style was to keep two blank lines?

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 5]
```diff
@@ -53,9 +52,16 @@ class VcfReadOptions(VariantTransformsOptions):
...
-                        required=True,
-                        help='Input pattern for VCF files to process.')
+    parser.add_argument(
+        '--input_pattern',
+        help=('Input pattern for VCF files to process. Either'
+              'this or --input_file flag has to be provided, exclusively.'))
+    parser.add_argument(
+        '--input_file',
+        help=('File that contains the list of VCF file names to input. Either '
+              'this or --input_pattern flag has to be provided, exlusively.'
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: typo (exlusively)

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 6]
```diff
@@ -53,9 +52,16 @@ class VcfReadOptions(VariantTransformsOptions):
 
   def add_arguments(self, parser):
     """Adds all options of this transform to parser."""
-    parser.add_argument('--input_pattern',
-                        required=True,
-                        help='Input pattern for VCF files to process.')
+    parser.add_argument(
+        '--input_pattern',
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: Can this be a remote file?

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 7]
```diff
@@ -53,9 +52,16 @@ class VcfReadOptions(VariantTransformsOptions):
...
+        help=('Input pattern for VCF files to process. Either'
+              'this or --input_file flag has to be provided, exclusively.'))
+    parser.add_argument(
+        '--input_file',
+        help=('File that contains the list of VCF file names to input. Either '
+              'this or --input_pattern flag has to be provided, exlusively.'
+              'Note that using input_file than input_pattern is slower for '
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: s/than/rather than/

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 8]
```diff
@@ -477,9 +475,16 @@ class PreprocessOptions(VariantTransformsOptions):
 
   def add_arguments(self, parser):
     # type: (argparse.ArgumentParser) -> None
-    parser.add_argument('--input_pattern',
-                        required=True,
-                        help='Input pattern for VCF files to process.')
+    parser.add_argument(
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: Unrelated to this CL but why is this duplicated?

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 9]
```diff
@@ -583,3 +592,45 @@ def add_arguments(self, parser):
               'be the same as the BigQuery table, but it requires all '
               'extracted variants to have the same call name ordering (usually '
               'true for tables from single VCF file import).'))
+
+
+def _validate_inputs(parsed_args):
+  if ((parsed_args.input_pattern and parsed_args.input_file) or
+      (not parsed_args.input_pattern and not parsed_args.input_file)):
+    raise ValueError('One and only one of input_pattern and input_file has '
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: Usually written as "Exactly one"

"Exactly one of input_pattern or input_file must be provided"

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 10]
```diff
@@ -583,3 +592,45 @@ def add_arguments(self, parser):
...
+                     'to be provided.')
+  if parsed_args.input_pattern is not None:
+    try:
+      # Gets at most 1 pattern match result of type `filesystems.MatchResult`.
+      first_match = filesystems.FileSystems.match(
```
Comment (kemp-google) [inline gcp_variant_transforms/options/variant_transform_options.py]: If there are no matches, does the array access not fail?

File: gcp_variant_transforms/pipeline_common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kemp-google) [reply gcp_variant_transforms/pipeline_common.py]: I was actually hoping we could do something like this for the flags themselves, but that will only work if we get a chance to read the file early in the process

File: gcp_variant_transforms/pipeline_common.py [hunk 11]
```diff
@@ -69,33 +69,62 @@ def parse_args(argv, command_line_options):
...
     return PipelineModes.LARGE
   elif total_files > _SMALL_DATA_THRESHOLD:
     return PipelineModes.MEDIUM
   return PipelineModes.SMALL
 
 
-def read_headers(pipeline, pipeline_mode, input_pattern):
+def get_file_names(input_file):
+  # type (str) -> List(str)
+  """ Reads all input file and extracts list of patterns out of it."""
```
Comment (kemp-google) [inline gcp_variant_transforms/pipeline_common.py]: All this does it takes a file and returns a list of the lines, correct?  Is there no python builtin for this?  (something like  open("foo").readlines() or so?)

File: gcp_variant_transforms/testing/integration/run_vcf_to_bq_tests.py [hunk 12]
```diff
@@ -16,7 +16,7 @@
 
 To define a new integration test case, create a json file in
 `gcp_variant_transforms/testing/integration/vcf_to_bq_tests` directory and
-specify at least test_name, table_name, and input_pattern for the integration
+specify at least test_name and table_name for the integration
```
Comment (kemp-google) [inline gcp_variant_transforms/testing/integration/run_vcf_to_bq_tests.py]: unrelated change?

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 13]
```diff
@@ -668,3 +593,45 @@ def add_arguments(self, parser):
...
               'extracted variants to have the same call name ordering (usually '
               'true for tables from single VCF file import).'))
+
+
+def _validate_inputs(parsed_args):
+  if ((parsed_args.input_pattern and parsed_args.input_file) or
+      (not parsed_args.input_pattern and not parsed_args.input_file)):
+    raise ValueError('Exactly one of input_pattern and input_file has to be '
+                     'provided.')
+  if parsed_args.input_pattern is not None:
```
Comment (allieychen) [inline gcp_variant_transforms/options/variant_transform_options.py]: Can be simplified as `if parsed_args.input_pattern:`

Comment (allieychen) [reply gcp_variant_transforms/options/variant_transform_options.py]: Since we also have similar validations in `_get_input_patterns` and `_get_file_names` (we have to validate the input pattern listed in the `input_file` as well), I prefer to move all validations except the first statement to `_get_file_names`. And the code can be simplified:
1. If `input_file` is provided, check whether it exists and read the contents and assign to known_args.input_pattern.
2. Otherwise, convert known_args.input_pattern to [known_args.input_patterns].
3. Now we can validate every item in input_pattern.

File: gcp_variant_transforms/pipeline_common.py [hunk 14]
```diff
@@ -66,82 +66,68 @@ def parse_args(argv, command_line_options):
...
+    for _, l in enumerate(f):
+      result.append(l)
+    if not result:
+      raise ValueError('Input file {} is empty.'.format(input_file))
+    return result
 
 def get_pipeline_mode(
-    input_pattern,
-    input_file,
+    input_patterns,
```
Comment (allieychen) [inline gcp_variant_transforms/pipeline_common.py]: Why we still have `input_file` at this point?

File: gcp_variant_transforms/pipeline_common.py [hunk 15]
```diff
@@ -66,82 +66,68 @@ def parse_args(argv, command_line_options):
...
   """Returns the mode the pipeline should operate in based on input size."""
   if optimize_for_large_inputs:
     return PipelineModes.LARGE
 
-  if input_pattern is not None:
-    total_files = _get_pipeline_mode_for_pattern_input(input_pattern)
-  else:
-    total_files = _get_pipeline_mode_for_file_input(input_file)
+  match_results = filesystems.FileSystems.match(input_patterns)
+  if not match_results:
```
Comment (allieychen) [inline gcp_variant_transforms/pipeline_common.py]: These validations should already be handled.

File: gcp_variant_transforms/testing/data/input_files/combine [hunk 16]
```diff
@@ -0,0 +1 @@
+gs://gcp-variant-transforms-testfiles/small_tests/valid-4.0*.vcf
```
Comment (allieychen) [inline gcp_variant_transforms/testing/data/input_files/combine]: It might be worthy more to have two file_patterns for the testing purpose.

File: gcp_variant_transforms/pipeline_common.py [hunk 17]
```diff
@@ -66,82 +66,68 @@ def parse_args(argv, command_line_options):
...
-      raise ValueError('Input file {} is empty.'.format(input_file))
-    return result
-
 
-def read_headers(pipeline, pipeline_mode, input_pattern, input_file):
+def read_headers(pipeline, pipeline_mode, input_patterns):
   # type: (beam.Pipeline, int, str) -> pvalue.PCollection
   """Creates an initial PCollection by reading the VCF file headers."""
-  if pipeline_mode == PipelineModes.LARGE or input_file:
+  if pipeline_mode == PipelineModes.LARGE or len(input_patterns) > 1:
```
Comment (allieychen) [inline gcp_variant_transforms/pipeline_common.py]: It might be easier to just change the pipeline_mode to LARGE in `get_pipeline_mode` when we have more than one input_pattern. But please check whether we use the pipeline mode anywhere else rather than reading the variants.

Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common.py]: Alas, annotation also looks up the pipeline mode, while it can have different input pattern (specifically, single input pattern of the shard directory).

File: gcp_variant_transforms/testing/data/input_files/combine (hunk 16 above)
Comment (tneymanov) [reply gcp_variant_transforms/testing/data/input_files/combine]: It does - there is * which encapsulates .../valid-4.0.vcf, .../valid-4.0-one-more-filed.vcf and .../valid-4.0-new.vcf.

File: gcp_variant_transforms/pipeline_common.py (hunk 15 above)
Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common.py]: So, as per offline convo, I tried to follow the older conventions and expand existing validations. Removed now.

File: gcp_variant_transforms/pipeline_common.py (hunk 14 above)
Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common.py]: Was needed for error message. Since we are removing validations from this section, removed.

File: gcp_variant_transforms/pipeline_common.py (hunk 17 above)
Comment (allieychen) [reply gcp_variant_transforms/pipeline_common.py]: The annotation pipeline also uses the pipeline_mode to read variants, I don't think it is different from other cases, right?

File: gcp_variant_transforms/testing/data/input_files/combine (hunk 16 above)
Comment (allieychen) [reply gcp_variant_transforms/testing/data/input_files/combine]: Sorry, what I mean is that we provide two patterns in this file, e.g., 
`gs://gcp-variant-transforms-testfiles/small_tests/valid-4.0*.vcf`
`gs://gcp-variant-transforms-testfiles/small_tests/valid-4.1*.vcf`

The new flag essentially allows the user to provide more than one pattern, while in this test case the user can actually use `input_pattern=gs://gcp-variant-transforms-testfiles/small_tests/valid-4.0*.vcf`

File: gcp_variant_transforms/testing/data/input_files/wrong [hunk 18]
```diff
@@ -0,0 +1,3 @@
+./gcp_variant_transforms/testing/data/vcf/valid-4.0.vcf
```
Comment (allieychen) [inline gcp_variant_transforms/testing/data/input_files/wrong]: It is better to include this in unit tests (i.e., create temporary files there).

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/platinum_NA12878_hg38_10K_lines_from_file.json [hunk 19]
```diff
@@ -0,0 +1,48 @@
+[
```
Comment (allieychen) [inline gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/platinum_NA12878_hg38_10K_lines_from_file.json]: It might be confusion, this test actually means the input file already contains the annotation information, and we would like to parse the annotation fields into different columns (by using `--annotation_fields` flag). `gnomad_genomes_GRCh37_chrX_head2500_run_vep` is one example that runs annotation pipelines.

File: gcp_variant_transforms/vcf_to_bq.py [hunk 20]
```diff
@@ -354,7 +366,9 @@ def run(argv=None):
 
   beam_pipeline_options = pipeline_options.PipelineOptions(pipeline_args)
   pipeline = beam.Pipeline(options=beam_pipeline_options)
-  variants = _read_variants(input_pattern, pipeline, known_args, pipeline_mode)
+  variants = _read_variants(input_patterns, pipeline, known_args, pipeline_mode)
+  logging.info('got here')
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: Please remove debugging info here.

File: gcp_variant_transforms/vcf_to_bq.py [hunk 21]
```diff
@@ -204,7 +209,7 @@ def _annotate_vcf_files(input_pattern, known_args, pipeline_args):
 
   with beam.Pipeline(options=options) as p:
     _ = (p
-         | beam.Create([input_pattern])
+         | beam.Create(input_patterns)
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: I think it would be useful to add/modify integration test for annotation, which runs with multiple input_patterns. You can modify the case `gnomad_genomes_GRCh37_chrX_head2500_run_vep` (by updating the input_pattern to input_file).

Comment (tneymanov) [reply gcp_variant_transforms/vcf_to_bq.py]: Done.

File: gcp_variant_transforms/vcf_to_bq.py (hunk 20 above)
Comment (tneymanov) [reply gcp_variant_transforms/vcf_to_bq.py]: Done.

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/platinum_NA12878_hg38_10K_lines_from_file.json (hunk 19 above)
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/platinum_NA12878_hg38_10K_lines_from_file.json]: Ahh added one more test for gnomad. Left plat test as well, just in case.

File: gcp_variant_transforms/testing/data/input_files/sample [hunk 22]
```diff
@@ -0,0 +1,3 @@
+./gcp_variant_transforms/testing/data/vcf/valid-4.0.vcf
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/data/input_files/sample]: Could you clarify what you mean here - do you want me to dynamically generate this and wrong files? Doesn't that seem like an overkill, or is there a precedence for this?

File: gcp_variant_transforms/pipeline_common.py (hunk 17 above)
Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common.py]: I'm sorry, I wasn't clear - at this [step](https://github.com/googlegenomics/gcp-variant-transforms/blob/master/gcp_variant_transforms/pipeline_common.py#L105) we potentially already have sharded the output so only have 1 input pattern, but whether input was large or not is still relevant

File: gcp_variant_transforms/testing/data/input_files/combine (hunk 16 above)
Comment (tneymanov) [reply gcp_variant_transforms/testing/data/input_files/combine]: Done.

File: gcp_variant_transforms/options/variant_transform_options.py (hunk 13 above)
Comment (tneymanov) [reply gcp_variant_transforms/options/variant_transform_options.py]: Done.

File: gcp_variant_transforms/testing/data/input_files/sample (hunk 22 above)
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/data/input_files/sample]: As per offline convo, Done.

File: gcp_variant_transforms/testing/data/input_files/wrong (hunk 18 above)
Comment (tneymanov) [reply gcp_variant_transforms/testing/data/input_files/wrong]: Done.

File: gcp_variant_transforms/testing/data/input_files/combine [hunk 23]
```diff
@@ -0,0 +1,2 @@
+gs://tural-test-runner/vcf_files/small/valid-4.1*
```
Comment (allieychen) [inline gcp_variant_transforms/testing/data/input_files/combine]: Why the files are from your project? Is this file the same as `gs://gcp-variant-transforms-testfiles/small_tests/input_files/combine_input`? If yes, please use the same file name.

File: gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json [hunk 24]
```diff
@@ -0,0 +1,14 @@
+[
+  {
+    "test_name": "no-conflicts",
+    "input_pattern": "gs://gcp-variant-transforms-testfiles/small_tests/input_files/combine_input",
```
Comment (allieychen) [inline gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json]: input_pattern? Or input_file?

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/medium_tests/gnomad_genomes_GRCh38_chrX_head2500_run_vep_from_file.json [hunk 25]
```diff
@@ -0,0 +1,86 @@
+[
+  {
+    "test_name": "gnomad-genomes-grch37-chr-x-head2500-run-vep",
+    "table_name": "gnomad_genomes_GRCh37_chrX_head2500_run_vep",
+    "input_pattern": "gs://gcp-variant-transforms-testfiles/small_tests/input_files/gnomad_input",
```
Comment (allieychen) [inline gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/medium_tests/gnomad_genomes_GRCh38_chrX_head2500_run_vep_from_file.json]: input_pattern?

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/platinum_NA12878_hg38_10K_lines_from_file.json [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+  {
```
Comment (allieychen) [inline gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/platinum_NA12878_hg38_10K_lines_from_file.json]: What's the purpose of this test?

File: gcp_variant_transforms/vcf_to_bq.py [hunk 27]
```diff
@@ -186,10 +190,10 @@ def _shard_variants(known_args, pipeline_args, pipeline_mode):
              beam.pvalue.AsSingleton(call_names),
              known_args.number_of_variants_per_shard))
 
-  return vep_runner_util.format_dir_path(vcf_shards_output_dir)
+  return [vep_runner_util.format_dir_path(vcf_shards_output_dir)]
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: Please update the return type reference accordingly.

File: gcp_variant_transforms/vcf_to_bq.py [hunk 28]
```diff
@@ -172,9 +176,9 @@ def _shard_variants(known_args, pipeline_args, pipeline_mode):
   _update_google_cloud_job_name(google_cloud_options, shard_files_job_name)
   vcf_shards_output_dir = filesystems.FileSystems.join(
       known_args.annotation_output_dir, _SHARDS_FOLDER)
+  input_patterns = known_args.input_patterns
   with beam.Pipeline(options=options) as p:
-    variants = _read_variants(known_args.input_pattern, p, known_args,
-                              pipeline_mode)
+    variants = _read_variants(input_patterns, p, known_args, pipeline_mode)
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: Why not use `known_args.input_patterns` directly?

File: gcp_variant_transforms/vcf_to_bq.py [hunk 29]
```diff
@@ -258,7 +262,7 @@ def _merge_headers(input_pattern, known_args, pipeline_args, pipeline_mode,
       temp_directory, temp_merged_headers_file_name)
 
   with beam.Pipeline(options=options) as p:
-    headers = pipeline_common.read_headers(p, pipeline_mode, input_pattern)
+    headers = pipeline_common.read_headers(p, pipeline_mode, input_patterns)
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: Why not use `known_args.input_patterns`?

File: gcp_variant_transforms/vcf_to_bq.py [hunk 30]
```diff
@@ -321,15 +328,19 @@ def run(argv=None):
 
   annotated_vcf_pattern = _run_annotation_pipeline(known_args, pipeline_args)
 
-  input_pattern = annotated_vcf_pattern or known_args.input_pattern
+  input_patterns = known_args.input_patterns
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: Why remove `annotated_vcf_pattern or`?

File: gcp_variant_transforms/testing/integration/run_preprocessor_tests.py [hunk 31]
```diff
@@ -71,8 +70,7 @@ def __init__(self,
     self._report_blob_name = self._append_suffix(report_blob_name, suffix)
     self._report_path = '/'.join(['gs:/', _BUCKET_NAME, self._report_blob_name])
     self._project = parser_args.project
-    args = ['--input_pattern {}'.format(input_pattern),
-            '--report_path {}'.format(self._report_path),
+    args = ['--report_path {}'.format(self._report_path),
```
Comment (allieychen) [inline gcp_variant_transforms/testing/integration/run_preprocessor_tests.py]: The input_pattern (or input_file) is require to run the test, Did you provide the input_pattern somewhere else?

File: gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json [hunk 32]
```diff
@@ -0,0 +1,14 @@
+[
+  {
+    "test_name": "no-conflicts",
```
Comment (allieychen) [inline gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json]: Please use a unique test_name.

Comment (allieychen) [reply gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json]: Since this test actually passes even when we wrongly provide `input_pattern`, so it is preferred to modify the input file to actually generate some conflicts information. For instance, you can add `gs://gcp-variant-transforms-testfiles/small_tests/invalid-4.0-POS-empty.vcf` into `combine`, the expected contents should be something similar to `report_malformed_records.json`

File: gcp_variant_transforms/vcf_to_bq.py [hunk 33]
```diff
@@ -85,7 +85,7 @@
 _SHARDS_FOLDER = 'shards'
 
 
-def _read_variants(input_pattern, pipeline, known_args, pipeline_mode):
+def _read_variants(all_patterns, pipeline, known_args, pipeline_mode):
   # type: (str, beam.Pipeline, argparse.Namespace, int) -> pvalue.PCollection
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: s/str/List[str]

File: gcp_variant_transforms/vcf_to_bq.py [hunk 34]
```diff
@@ -131,7 +132,7 @@ def _get_variant_merge_strategy(known_args  # type: argparse.Namespace
     raise ValueError('Merge strategy is not supported.')
 
 
-def _add_inferred_headers(input_pattern,  # type: str
+def _add_inferred_headers(all_patterns,  # type: str
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: s/str/List[str]

File: gcp_variant_transforms/vcf_to_bq.py [hunk 35]
```diff
@@ -186,10 +190,10 @@ def _shard_variants(known_args, pipeline_args, pipeline_mode):
              beam.pvalue.AsSingleton(call_names),
              known_args.number_of_variants_per_shard))
 
-  return vep_runner_util.format_dir_path(vcf_shards_output_dir)
+  return [vep_runner_util.format_dir_path(vcf_shards_output_dir)]
 
 
-def _annotate_vcf_files(input_pattern, known_args, pipeline_args):
+def _annotate_vcf_files(all_patterns, known_args, pipeline_args):
   # type: (str, argparse.Namespace, List[str]) -> str
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: s/str/List[str]

File: gcp_variant_transforms/vcf_to_bq.py [hunk 36]
```diff
@@ -229,13 +233,12 @@ def _update_google_cloud_job_name(google_cloud_options, job_name):
     google_cloud_options.job_name = job_name
 
 
-def _merge_headers(input_pattern, known_args, pipeline_args, pipeline_mode,
-                   annotated_vcf_pattern=None):
+def _merge_headers(known_args, pipeline_args,
+                   pipeline_mode, annotated_vcf_pattern=None):
   # type: (str, argparse.Namespace, List[str], int, str) -> None
```
Comment (allieychen) [inline gcp_variant_transforms/vcf_to_bq.py]: Please update the type.

File: gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json (hunk 32 above)
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/preprocessor_tests/no_conflicts_from_file.json]: Done.

File: gcp_variant_transforms/vcf_to_bq.py (hunk 34 above)
Comment (tneymanov) [reply gcp_variant_transforms/vcf_to_bq.py]: Done.

File: gcp_variant_transforms/pipeline_common.py [hunk 37]
```diff
@@ -87,15 +131,16 @@ def get_pipeline_mode(input_pattern, optimize_for_large_inputs=False):
...
+  # type: (beam.Pipeline, int, List[str]) -> pvalue.PCollection
   """Creates an initial PCollection by reading the VCF file headers."""
   if pipeline_mode == PipelineModes.LARGE:
     headers = (pipeline
-               | beam.Create([input_pattern])
+               | beam.Create(all_patterns)
                | vcf_header_io.ReadAllVcfHeaders())
   else:
-    headers = pipeline | vcf_header_io.ReadVcfHeaders(input_pattern)
+    headers = pipeline | vcf_header_io.ReadVcfHeaders(all_patterns[0])
```
Comment (kemp-google) [inline gcp_variant_transforms/pipeline_common.py]: Is this safe?  What if all_patterns is empty?  Are you just counting on this not being called unless there is at least one pattern in the list?

Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common.py]: It should be - we actually had this discussion before hand with regards to having validation on multiple places or only at the beginning and decided that it probably isn't worth to copy an unreachable code all over the pipeline, so long as the validation in the beginning is done right. So yeah, I'm counting that all_patterns has been verified until this point and every pattern in all patterns has at least 1 match in the filesystem
